web/packages/teleterm/src/ui/services/workspacesService/workspacesService.ts

```diff
@@ -342,11 +341,16 @@ export class WorkspacesService extends ImmutableStore<WorkspacesState> {
     const omitUriAndTitle = (documents: Document[]) =>
       documents.map(d => ({ ...d, uri: undefined, title: undefined }));
 
+    if (!previousDocuments) {
+      return false;
+    }
+
+    const omittedPreviousDocs = omitUriAndTitle(previousDocuments);
+    const omittedCurrentDocs = omitUriAndTitle(currentDocuments);
     return (
-      previousDocuments?.length &&
-      !isEqual(
-        omitUriAndTitle(previousDocuments),
-        omitUriAndTitle(currentDocuments)
+      // Ensure they are equal
+      !Object.keys(omittedPreviousDocs).some(
+        key => omittedPreviousDocs[key] === omittedCurrentDocs[key]
```

Comment:
The `_.isEqual` is quite extensive as it compares pretty much every data type. After talking with @ravicious these could just be shallowly compared.

Reply:
This check is incorrect 😬
If at least one value is the same in both objects it returns false.
I'd use `every`:
```ts
Object.keys(omittedPreviousDocs).every(
     key => omittedPreviousDocs[key] === omittedCurrentDocs[key]
);
```

Reply:
Oops! Good catch, I was using `some` so that it would short-circuit if there was a failure...no need to run the whole object if the first fails.

that was supposed to be:

```js
      !Object.keys(omittedPreviousDocs).some(
        key => omittedPreviousDocs[key] !== omittedCurrentDocs[key]
      )
```

Reply:
> I was using some so that it would short-circuit if there was a failure

you described `every` too :D 
>It calls a provided callbackFn function once for each element in an array, until the callbackFn returns a [falsy](https://developer.mozilla.org/en-US/docs/Glossary/Falsy) value. If such an element is found, every() immediately returns false and stops iterating through the array.

And I just realized that shallow comparison won't work here. We are comparing an array of objects we read from disk with an array of objects that we just created in JS, so they are never equal.

> After talking with @ravicious these could just be shallowly compared.

Did I miss something @ravicious?

Reply:
I mean the individual docs themselves can be shallowly compared, [they don't contain any scalar fields](https://github.com/gravitational/teleport/blob/master/web/packages/teleterm/src/ui/services/workspacesService/documentsService/types.ts). But the arrays themselves can't, as is the case in JavaScript.

`omittedPreviousDocs` and `omittedCurrentDocs`, `Object.keys` returns array indices when called on those vars. We want to check if both arrays contain objects with the same properties in the same order.

Reply:
Oops, I mean they contain only scalar fields.

Reply:
I believe this should work: https://stackoverflow.com/a/55256318/742872

Reply:
Sorry for the confusion folks, this has now been updated with a more robust comparison utility.